### PR TITLE
Deeper Instrumentation for Hybrid Retrievers + examples

### PR DESCRIPTION
**Pull request by joshreini1**
addresses: https://github.com/truera/trulens/issues/854 +slack requests

Instrumentation updates:

- Add instrumentation for postprocessor and retriever, query_engine components
- Recursively instrument sub-components. This is critical for hybrid/ensemble retrievers.

This results in much better instrumentation (and evals!). Previously, none of the subcomponents of `RetrieverQueryEngine.retrieve` were captured.

Screenshot 2024-02-08 at 7 10 16 PM

Examples:
Add user requested example for evaluating ensemble retrievers (langchain) and hybrid (llama-index)

Cool result:
Screenshot 2024-02-07 at 6 07 02 PM

### CODE REVIEW

**Comment by joshreini1** [comment]
Note: this does slow down the wrap application step. Trade-off seems worth it to get better instrumentation.

File: trulens_eval/trulens_eval/instruments.py [hunk 1]
```diff
@@ -743,6 +743,13 @@ def instrument_object(
         # https://github.com/pydantic/pydantic/blob/11079e7e9c458c610860a5776dc398a4764d538d/pydantic/main.py#LL370C13-L370C13
         # .
 
+        # Recursively instrument inner components
```
Comment (piotrm0) [inline trulens_eval/trulens_eval/instruments.py:746]: This should already be done by what is on line 835. Do you have example of that not recurring in correctly? If so, can you fix it there?

Comment (joshreini1) [reply trulens_eval/trulens_eval/instruments.py:746]: The classes are already included in instrumentation list in tru_llama. What do you mean by "fix it there"?

Comment (piotrm0) [reply trulens_eval/trulens_eval/instruments.py:746]: I mean the subcomponent instrumentation is done around line 835 so I was suggesting that your fix to it not correctly recurring down should be put in the same place.

**Comment by joshreini1** [comment]
User validation that this change solves the issue: https://github.com/truera/trulens/issues/854

File: trulens_eval/trulens_eval/instruments.py (hunk 1 above)
Comment (joshreini1) [reply trulens_eval/trulens_eval/instruments.py:745]: L835 seems to cover unrecognized container types? I don't think that's what's happening here

Comment (joshreini1) [reply trulens_eval/trulens_eval/instruments.py:745]: Would like to merge this soon, got customer ask for this today.

**Review by piotrm0 — APPROVED** [review]

**Pull request merged**